// multiply/esm/hql.js
import * as dntShim from "./_dnt.shims.js";
import { Env } from "./modules/env.js";
import { repl } from "./modules/repl.js";
import { compileHQL } from "./modules/compiler.js";
import { buildImportMap, buildImportMapForJS } from "./modules/importMap.js";
import { join, dirname, basename, extname, isAbsolute, resolve, cwd, readTextFile, mkdir, writeTextFile, execPath, run } from "./platform/platform.js";
async function startRepl() {
    const env = new Env();
    await repl(env);
}
async function start(args) {
    if (args.length < 2) {
        console.log("Usage:");
        console.log("  hql run <file>");
        dntShim.Deno.exit(1);
    }
    const file = args[1];
    const projectRoot = cwd();
    const cacheDir = join(projectRoot, ".hqlcache");
    const importMap = { imports: {} };
    let entryFile = file;
    if (file.endsWith(".hql")) {
        const absoluteInput = resolve(file);
        const source = await readTextFile(absoluteInput);
        const compiled = await compileHQL(source, absoluteInput, true);
        const outputFolder = absoluteInput.includes("/test/")
            ? join(dirname(absoluteInput), "transpiled")
            : dirname(absoluteInput);
        await mkdir(outputFolder, { recursive: true });
        const baseName = basename(absoluteInput, extname(absoluteInput));
        const outputFile = join(outputFolder, `${baseName}.hql.js`);
        await writeTextFile(outputFile, compiled);
        entryFile = outputFile;
        importMap.imports = await buildImportMap(absoluteInput, cacheDir);
    }
    else {
        entryFile = resolve(file);
        importMap.imports = await buildImportMapForJS(entryFile, cacheDir);
    }
    if (Object.keys(importMap.imports).length > 0) {
        const importMapPath = join(projectRoot, "hql_import_map.json");
        await writeTextFile(importMapPath, JSON.stringify(importMap, null, 2));
        const command = [
            execPath(),
            "run",
            `--import-map=${importMapPath}`,
            "--allow-read",
            "--allow-write",
            "--allow-net",
            "--allow-env",
            "--allow-run",
            entryFile,
        ];
        await execute(command);
    }
    else {
        const command = [
            execPath(),
            "run",
            "--allow-read",
            "--allow-write",
            "--allow-net",
            "--allow-env",
            "--allow-run",
            entryFile,
        ];
        await execute(command);
    }
}
async function execute(cmd) {
    const proc = run(cmd);
    const status = await proc.status();
    dntShim.Deno.exit(status.code);
}
async function transpile(args) {
    if (args.length < 2) {
        console.log("Usage:");
        console.log("  hql transpile <inputFile> [outputFile]");
        dntShim.Deno.exit(1);
    }
    const inputFile = args[1];
    const absoluteInput = resolve(inputFile);
    const projectRoot = cwd();
    let outputFile;
    if (args.length >= 3) {
        outputFile = args[2];
        if (!isAbsolute(outputFile)) {
            outputFile = resolve(join(projectRoot, outputFile));
        }
    }
    else {
        const baseName = basename(absoluteInput, extname(absoluteInput));
        outputFile = join(dirname(absoluteInput), `${baseName}.hql.js`);
    }
    const source = await readTextFile(absoluteInput);
    const compiled = await compileHQL(source, absoluteInput, false);
    await mkdir(dirname(outputFile), { recursive: true });
    await writeTextFile(outputFile, compiled);
    console.log(`Transpiled ${absoluteInput} -> ${outputFile}`);
}
async function main() {
    const args = dntShim.Deno.args;
    if (args.length === 0) {
        await startRepl();
        return;
    }
    const command = args[0];
    switch (command) {
        case "repl":
            await startRepl();
            break;
        case "run":
            await start(args);
            break;
        case "transpile":
            await transpile(args);
            break;
        default:
            console.log("Unknown command.");
            console.log("Usage:");
            console.log("  hql repl");
            console.log("  hql run <file>");
            console.log("  hql transpile <inputFile> [outputFile]");
            dntShim.Deno.exit(1);
    }
}
if (import.meta.main) {
    main();
}
export { exportHqlModules, getHqlModule } from "./modules/export.js";
