// multiply/esm/platform/platform.js
// // platform/platform.ts
// const isDeno = "Deno" in globalThis && globalThis.Deno?.version != null;
// let platformModule: any;
// if (isDeno) {
//   platformModule = await import("./deno/platform.ts");
// } else {
//   const nodeModulePath = "./node/" + "platform.ts";
//   platformModule = await import(nodeModulePath);
// }
// export const cwd = platformModule.cwd;
// export const stat = platformModule.stat;
// export const readTextFile = platformModule.readTextFile;
// export const writeTextFile = platformModule.writeTextFile;
// export const mkdir = platformModule.mkdir;
// export const join = platformModule.join;
// export const dirname = platformModule.dirname;
// export const basename = platformModule.basename;
// export const extname = platformModule.extname;
// export const isAbsolute = platformModule.isAbsolute;
// export const resolve = platformModule.resolve;
// export const relative = platformModule.relative;
// export const realPathSync = platformModule.realPathSync;
// export const execPath = platformModule.execPath;
// export const run = platformModule.run;
// Deno-specific code
import * as dntShim from "../_dnt.shims.js";
import { join as stdJoin, dirname as stdDirname, basename as stdBasename, extname as stdExtname, isAbsolute as stdIsAbsolute, resolve as stdResolve, relative as stdRelative, } from "../deps/deno.land/std@0.170.0/path/mod.js";
export function cwd() {
    return dntShim.Deno.cwd();
}
export async function stat(path) {
    return await dntShim.Deno.stat(path);
}
export async function readTextFile(path) {
    return await dntShim.Deno.readTextFile(path);
}
export async function writeTextFile(path, data) {
    return await dntShim.Deno.writeTextFile(path, data);
}
export async function mkdir(path, opts) {
    return await dntShim.Deno.mkdir(path, opts);
}
export function join(...segments) {
    return stdJoin(...segments);
}
export function dirname(path) {
    return stdDirname(path);
}
export function basename(path, ext) {
    return stdBasename(path, ext);
}
export function extname(path) {
    return stdExtname(path);
}
export function isAbsolute(path) {
    return stdIsAbsolute(path);
}
export function resolve(...segments) {
    return stdResolve(...segments);
}
export function relative(from, to) {
    return stdRelative(from, to);
}
export function realPathSync(path) {
    return dntShim.Deno.realPathSync(path);
}
export function execPath() {
    return dntShim.Deno.execPath();
}
export function run(cmd) {
    return dntShim.Deno.run({ cmd });
}
